app/src/main/java/org/kiwix/kiwixmobile/main/AddNoteDialog.java

```diff
@@ -0,0 +1,427 @@
+      MenuItem saveItem = toolbar.getMenu().findItem(R.id.save_note);
+      saveItem.setEnabled(true);
+      saveItem.getIcon().setAlpha(255);
+
+    } else {
+      Log.d(TAG, "Toolbar without inflated menu");
+    }
+  }
+
+  private void enableShareNoteMenuItem() {
+    if(toolbar.getMenu() != null) {
+      MenuItem shareItem = toolbar.getMenu().findItem(R.id.share_note);
+      shareItem.setEnabled(true);
+      shareItem.getIcon().setAlpha(255);
+
+    } else {
+      Log.d(TAG, "Toolbar without inflated menu");
+    }
+  }
+
+  @Override
+  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+    super.onViewCreated(view, savedInstanceState);
+
+    if(!noteFileExists) {
+      // Prepare for input in case of empty/new note
+      addNoteEditText.requestFocus();
+      showKeyboard();
+    }
+  }
+
+  public void showKeyboard(){
+    InputMethodManager inputMethodManager = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
+    inputMethodManager.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);
+  }
+
+  public void closeKeyboard(){
+    InputMethodManager inputMethodManager = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
+    inputMethodManager.toggleSoftInput(InputMethodManager.HIDE_IMPLICIT_ONLY, 0);
+  }
+
+  private void saveNote(String noteText) {
+
+    /* String content of the EditText, given by noteText, is saved into the text file given by:
+     *    "{External Storage}/Kiwix/Notes/ZimFileTitle/ArticleTitle.txt"
+     * */
+
+    if(isExternalStorageWritable()) {
+
+      if(ContextCompat.checkSelfPermission(getContext(), Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
+        Log.d(TAG, "WRITE_EXTERNAL_STORAGE permission not granted");
+        Toast.makeText(getContext(), getString(R.string.note_save_unsuccessful), Toast.LENGTH_LONG);
+        return;
+      }
+
+      File notesFolder = new File(Environment.getExternalStorageDirectory() + "/Kiwix/Notes/" + zimFileTitle);
+      boolean folderExists = true;
+
+      if(!notesFolder.exists()) {
+        // Try creating folder if it doesn't exist
+        folderExists = notesFolder.mkdirs();
+      }
+
+      if(folderExists) {
+        File noteFile = new File(notesFolder.getAbsolutePath(), articleTitle + ".txt");
+
+        // Save note text-file code:
+        try {
+          FileOutputStream fileOutputStream = new FileOutputStream(noteFile);
+          fileOutputStream.write(noteText.getBytes());
+          fileOutputStream.close();
+          Toast.makeText(getContext(), getActivity().getString(R.string.note_save_successful), Toast.LENGTH_SHORT).show();
+          noteEdited = false; // As no unsaved changes remain
+
+        } catch (IOException e) {
+          e.printStackTrace();
+          Toast.makeText(getContext(), getActivity().getString(R.string.note_save_unsuccessful), Toast.LENGTH_LONG);
```

Comment:
no .show() here and on the next one. Android studio should generate a warning about this, if it is not you should review your inspections in the settings.
Extract a common method (place cursor on method crtl+alt+m) for all these toast methods so it is impossible to make this mistake. Also there exists Toast.makeText(context int, int) where the second int accepts a string reference removing the need for a getString call.

Reply:
Corrected [here](https://github.com/kiwix/kiwix-android/pull/1198/commits/873eb2bce86c76b11962367701cb66bdf66b9f5e#diff-36e13fa4a9132616aee81e1f88bfa0ffR258)

Since I am using length_long for errors and short for success-confirmations, an extracted method will need 2 parameters again (string and duration). It's one less than the .makeText(), so I didn't refactor to an extra method at the moment. But I'll definitely use an extracted method in the future

Also I enabled all the previously disabled 'android' inspection options as you suggested, but it still didn't flag the .show()-less line. Will dig into it.

Reply:
![image](https://user-images.githubusercontent.com/3358282/59595942-b2ecf280-90ee-11e9-8ebe-71c45dc8fbd1.png)
the extracted method is `showToast(int stringId, int duration)` it combines the functionality of make & .show() preventing an error from happening.
This comment from me was more to try to get you to see the importance of the DRY principle and also if you were unaware of the incredibly useful extract keyboard shortcuts then to highlight those. They will automatically convert all usages in the file if you do it correctly, highlight one Toast.makeText().show() and press ctr+alt+m (this is on windows, if your machine is differenct right click>refactor>extract>method), if at this point it doesn't auto suggest extracting the parameters too then do it manually, in this new method extract the stringId to a parameter by pressing ctrl+alt+p, do the same for the duration, now almost certainly all the code will be replaced for you.
This is mostly an academic exercise, come the kotlin introduction there is an extension method on context called toast() that has default arguments for duration

Reply:
Refactored [here](https://github.com/kiwix/kiwix-android/pull/1198/commits/ffba82e0c79d0c539d55c336a9ae43416a71674e)

Thank you for informing about DRY - I developed a habit of avoiding functions if possible as I was used to solving coding problems more than doing development work (in-lining is never assured; performance hit; etc). I should revise the OOP topics again